src/pg_to_evalscript/javascript_datacube/DataCube.js

```diff
@@ -228,6 +228,91 @@ class DataCube {
         dimension.labels = labels && labels.length > 0 ? labels : computedLabels;
     }
 
+    aggregateTemporalPeriod(period, reducer, dimension, context) {
+        const temporalDimensions = this.getTemporalDimensions();
+        if (!dimension && temporalDimensions.length > 1) {
+          throw new ProcessError({
+            name: "TooManyDimensions",
+            message:
+              "The data cube contains multiple temporal dimensions. The parameter `dimension` must be specified.",
+          });
+        }
+      
+        const temporalDimensionToAggregate = dimension ? this.getDimensionByName(dimension) : temporalDimensions[0];
+        if (!temporalDimensionToAggregate) {
+          throw new ProcessError({
+            name: "DimensionNotAvailable",
+            message: "A dimension with the specified name does not exist.",
+          });
+        }
+      
+        const axis = this.dimensions.findIndex(
+          (d) => (d.name = temporalDimensionToAggregate.name)
+        );
+        const newLabels = [];
+        const newValues = [];
+      
+        if (temporalDimensionToAggregate.labels.length > 1) {
+          const { minDate, maxDate } = getMinMaxDate(
+            temporalDimensionToAggregate.labels
+          );
+          const firstDayInYear = new Date(minDate)
+          firstDayInYear.setMonth(0)
+          firstDayInYear.setDate(1)
+      
+          const dates = generateDatesInRangeByPeriod(firstDayInYear.toISOString(), maxDate, period);
+          let shouldAdd = false;
+          for (let i = 0; i < dates.length; i++) {
+              if (formatLabelByPeriod(period, dates[i]) === formatLabelByPeriod(period,minDate)) {
+                shouldAdd = true;
+              }  
+
+              if (shouldAdd) {
+                newLabels.push(formatLabelByPeriod(period, dates[i]));
+              }
+          }
+        } else {
+          newLabels.push(
+            formatLabelByPeriod(period, temporalDimensionToAggregate.labels[0])
+          );
+        }
+
+        const formattedOldLabels = temporalDimensionToAggregate.labels.map(l => formatLabelByPeriod(period, l))
+        for (let newLabel of newLabels) {
+          const allCoords = this._iterateCoords(this.data.shape.slice(), [axis]);
+          const indices = []
+          for (let i = 0; i < formattedOldLabels.length; i++) {
+              if (newLabel === formattedOldLabels[i]) {
```

Comment:
Not really sure if I understand this correctly. We first compile a list of `newLabels` which is what the final labels should look like. In what way does `formattedOldLabels` differ from `newLabels` that we are comparing them here and in what way are they the same?

Could this be somehow merged with the `for` loop [here](https://github.com/openEOPlatform/openeo-pg-evalscript-converter/pull/67/files#diff-79a8177999eff94809eb3e5655bc2ff8f331addad600ca4a7f7ca6bbe53297ebR264)?

Reply:
Yes `newLabels` is what the final formatted labels should look like. While `formattedOldLabels` are basically just that, formatted. E. g. if we had `period="day"` and `labels= ["2020-01-01T00:00:00:123Z", "2020-01-04T03:23:15:345Z", "2020-01-04T12:45:23:634Z"]`, this would produce `newLabels = ["2020-001", "2020-002", "2020-003", "2020-004"]`, while formatted old labels would look like this `formattedOldLabels = ["2020-001", "2020-004", "2020-004"]`. And by looking at these 2 arrays, I can figure out on which label there is available data and which labels contain no data and which labels should be aggregated.

It can be merged yes, moved it in that for loop that you mentioned, but also had to move it to the else statement below this for loop, so I dont know if this is actually an improvement, since now this code is duplicated

Reply:
Thanks for the explanation, that makes sense. 

> It can be merged yes, moved it in that for loop that you mentioned, but also had to move it to the else statement below this for loop, so I dont know if this is actually an improvement, since now this code is duplicated

Not sure this is what I had in mind. You can also put it back to not have this duplication (or extract it into a single function).